processing/src/main/java/io/druid/query/lookup/LookupReferencesManager.java

```diff
@@ -52,333 +55,400 @@
-  private final AtomicBoolean started = new AtomicBoolean(false);
+  private static final EmittingLogger LOG = new EmittingLogger(LookupReferencesManager.class);
+
+  // Lookups state (loaded/to-be-loaded/to-be-dropped etc) is managed by immutable LookupUpdateState instance.
+  // Any update to state is done by creating updated LookupUpdateState instance and atomically setting that
+  // into the ref here.
+  // this allows getAllLookupsState() to provide a consistent view without using locks.
+  @VisibleForTesting
+  final AtomicReference<LookupUpdateState> stateRef = new AtomicReference<>();
+
+  @VisibleForTesting
+  final LookupSnapshotTaker lookupSnapshotTaker;
+
+  @VisibleForTesting
+  final LifecycleLock lifecycleLock = new LifecycleLock();
+
+  @VisibleForTesting
+  Thread mainThread;
 
-  private final LookupSnapshotTaker lookupSnapshotTaker;
+  //for unit testing only
+  private final boolean testMode;
 
   @Inject
-  public LookupReferencesManager(LookupConfig lookupConfig, final @Json ObjectMapper objectMapper)
+  public LookupReferencesManager(LookupConfig lookupConfig, @Json ObjectMapper objectMapper)
+  {
+    this(lookupConfig, objectMapper, false);
+  }
+
+  @VisibleForTesting
+  LookupReferencesManager(LookupConfig lookupConfig, ObjectMapper objectMapper, boolean testMode)
   {
     if (Strings.isNullOrEmpty(lookupConfig.getSnapshotWorkingDir())) {
       this.lookupSnapshotTaker = null;
     } else {
       this.lookupSnapshotTaker = new LookupSnapshotTaker(objectMapper, lookupConfig.getSnapshotWorkingDir());
     }
+
+    this.testMode = testMode;
   }
 
   @LifecycleStart
   public void start()
   {
-    startStopLock.writeLock().lock();
+    if (!lifecycleLock.canStart()) {
+      throw new ISE("can't start.");
+    }
+
     try {
-      if (!started.getAndSet(true)) {
-        if (lookupSnapshotTaker != null) {
-          final List<LookupBean> lookupBeanList = lookupSnapshotTaker.pullExistingSnapshot();
-          for (LookupBean lookupBean : lookupBeanList) {
-            this.put(lookupBean.name, lookupBean.factory);
-          }
-        }
-        LOGGER.info("Started lookup factory references manager");
+      LOG.info("LookupReferencesManager is starting.");
+
+      loadSnapshotAndInitStateRef();
+
+      if (!testMode) {
+        mainThread = Execs.makeThread(
+            "LookupReferencesManager-MainThread",
+            new Runnable()
+            {
+              @Override
+              public void run()
+              {
+                try {
+
+                  if (!lifecycleLock.awaitStarted()) {
+                    LOG.error("WTF! lifecycle not started, lookup update notices will not be handled.");
+                    return;
+                  }
+
+                  while (!Thread.interrupted() && lifecycleLock.awaitStarted(1, TimeUnit.MILLISECONDS)) {
+                    try {
+                      handlePendingNotices();
```

Comment:
(I might be misreading code paths but here it goes)

Here's what I see happening:

1. A lookup factory starts up and tries to fetch something from a remote system
2. The lookup factory's fetcher method catches InterruptedException (thus clearing the interrupt flag on the thread)
3. The lookup factory wraps the exception in a RuntimeException and bubbles it up
4. The loop goes around and doesn't see an interrupt flag on the current thread because it was cleared by catching the InterruptedException
5. Loop continues.